uaclient/api/u/pro/detach/v1.py

```diff
@@ -0,0 +1,100 @@
+from typing import List
+
+from uaclient import daemon, entitlements, exceptions, messages, timer, util
+from uaclient.api.api import APIEndpoint
+from uaclient.api.data_types import AdditionalInfo, ErrorWarningObject
+from uaclient.api.u.pro.security.status.reboot_required.v1 import (
+    _reboot_required,
+)
+from uaclient.api.u.pro.status.is_attached.v1 import _is_attached
+from uaclient.config import UAConfig
+from uaclient.data_types import (
+    BoolDataValue,
+    DataObject,
+    Field,
+    StringDataValue,
+    data_list,
+)
+from uaclient.files import state_files
+from uaclient.timer.update_messaging import update_motd_messages
+
+
+class DetachResult(DataObject, AdditionalInfo):
+    fields = [
+        Field("disabled", data_list(StringDataValue)),
+        Field("reboot_required", BoolDataValue),
+    ]
+
+    def __init__(self, disabled: List[str], reboot_required: bool):
+        self.disabled = disabled
+        self.reboot_required = reboot_required
+
+
+def detach() -> DetachResult:
+    return _detach(UAConfig())
+
+
+def _detach(cfg: UAConfig) -> DetachResult:
+    if not util.we_are_currently_root():
+        raise exceptions.NonRootUserError
+
+    if not _is_attached(cfg).is_attached:
+        raise exceptions.UnattachedError()
```

Comment:
I think we should have an early return success here with an empty disabled list - similar to how we decided "attach" should act if already attached.

Reply:
Maybe I am overthinking this, but in the situation where desktop consumes that API, wouldn't it be better to tell the user that the machine is not attached than always saying to the user that detach "worked".

But let me know if this is overthinking

Reply:
We have agreed that this should not be an error. I will update the PR to reflect that